src/hardware/adc_internal.rs

```diff
@@ -0,0 +1,133 @@
+use super::hal::{
+    adc,
+    gpio::{gpioa::*, gpiob::*, gpioc::*, gpiof::*, Analog},
+    hal::blocking::delay::DelayUs,
+    prelude::*,
+    rcc::{rec, CoreClocks},
+    stm32::{ADC1, ADC2, ADC3},
+};
+
+pub enum TecChannel {
+    Zero,
+    One,
+    Two,
+    Three,
+}
+
+pub enum Supply {
+    P5v,
+    P12v,
+    P3v3,
+    I12v,
+}
+
+pub struct AdcInternal {
+    adc1: adc::Adc<ADC1, adc::Enabled>,
+    adc3: adc::Adc<ADC3, adc::Enabled>,
+    p5v: PC0<Analog>,
+    p12v: PC2<Analog>,
+    p3v3: PF7<Analog>,
+    i12v: PF8<Analog>,
+    tecu: TecUPins,
+    teci: TecIPins,
+}
+
+const V_REF: f32 = 3.0; // ADC reference voltage
+const DIV_P12V: f32 = 1.6 / (1.6 + 6.8); // Resistor divider 12V rail
+const DIV_P5V: f32 = 6.8 / (10.0 + 6.8); // Resistor divider 5V rail
+const DIV_P3V3: f32 = 6.8 / (1.6 + 6.8); // Resistor divider 3V3 rail
+const CONF_I12V: f32 = 0.005 * (10000.0 / 100.0); // 12V current measurement resistor configuration for LT6106
+
+pub type TecUPins = (PC3<Analog>, PA0<Analog>, PA3<Analog>, PA4<Analog>);
+pub type TecIPins = (PA5<Analog>, PA6<Analog>, PB0<Analog>, PB1<Analog>);
+
+impl AdcInternal {
+    #![allow(clippy::too_many_arguments)]
```

Comment:
clippy is right.
Maybe `SupplyPins` is enough here.
If that's not enough, do inline tuples:

```rust
pub fn new(... adc: (ADC1, ADC2, ADC3), ...)
```